File: solutions/leetcode/subsets.cpp
#include <vector>

class Solution {
 public:
  std::vector<std::vector<int>> subsets(std::vector<int> const& nums, size_t i = 0) {
    if (i < nums.size()) {
      std::vector<std::vector<int>> ans{subsets(nums, i + 1)};
      size_t const s = ans.size();
      for (size_t j = 0; j < s; ++j) {
        std::vector<int> v{ans[j]};
        v.push_back(nums[nums.size() - 1 - i]);
        ans.push_back(std::move(v));
      }
      return ans;
    } else {
      return {{}};
    }
  }
};

File: solutions/leetcode/subsets.rs
/* Problem subsets
 * https://www.leetcode.com/problems/subsets
 */

struct Solution;

impl Solution {
	pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
		nums.into_iter().fold(vec![vec![]], |mut subsets, num| {
			for i in 0..subsets.len() {
				let mut subset = subsets[i].clone();
				subset.push(num);
				subsets.push(subset);
			}
			subsets
		})
	}
}

fn main() {}
